proposals/mpi-operator-proposal.md

```diff
@@ -0,0 +1,224 @@
+## Motivation
+Kubeflow currently supports distributed training of TensorFlow models using
+[tf-operator](https://github.com/kubeflow/tf-operator), which relies on
+centralized parameter servers for coordination between workers. An alternative
+is a decentralized approach in which workers communicate with each other
+directly via the MPI `allreduce` primitive, without using parameter servers.
+NVIDIA's [NCCL](https://developer.nvidia.com/nccl) library has implemented most
+of the MPI primitives efficiently on GPUs, and Uber's
+[Horovod](https://github.com/uber/horovod) makes it simple to do multi-gpu and
+multi-node training using TensorFlow; the TensorFlow team is also working on an
+`allreduce`
+[implementation](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/contrib/distribute).
+Compared to parameter servers, this approach arguably is more
+[bandwidth optimal](http://www.cs.fsu.edu/~xyuan/paper/09jpdc.pdf) and
+[scales better](https://eng.uber.com/horovod/).
+
+However, right now there is no easy way to launch this type of training jobs on
+Kubernetes. By providing a CRD and a custom controller, we can make
+allreduce-style distributed training as simple as training on a single node.
+
+## Goals
+* Provide a common Custom Resource Definition (CRD) for defining a single-gpu,
+multi-gpu, or multi-node training job.
+* Implement a custom controller to manage the CRD, create dependent resources,
+and reconcile the desired states.
+* The cross-pod communication should be secure, without granting unnecessary
+permissions to any pod.
+* Though the initial version focuses on TensorFlow/Horovod, the approach can be
+applied to other frameworks with MPI support, such as CNTK.
+
+## Non-Goals
+In theory, this operator can be used to run arbitrary MPI jobs (e.g.
+computational fluid dynamics), but it's not our focus.
+
+## API
+
+### Container Image
+
+A requirement on user's container image is that it includes an MPI
+implementation. The initial version is tested on Open MPI 3.0.0, but it can be
+extended to other MPI implementations if there is a need.
```

Comment:
Could someone from the community comment on why Open MPI < 3.x would be valuable? I am not familiar.

Reply:
@galv I think yes.  OpenMPI 3.0.x has a bug on GPUDirect  (I actually use this feature).

ref: https://github.com/open-mpi/ompi/issues/3972

NOTE:  infiniband cannot be scheduled by default, but it can be used via very thin device plugin which just exposes device files in a host to the containers.

Reply:
oh, I noticed OpenMPI v3.0.2(not released yet) will fix the issue !! 👍